src/Chapter2.hs

```diff
@@ -858,7 +893,16 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [Int] -> [Int]
+rewind [] = []
+-- rewind (x: xs) = rewind xs ++ [x]
+rewind (x: xs) = rewind2 xs [x]
+  where
+    rewind2 :: [Int] -> [Int] -> [Int]
+    rewind2 [] l = l
+    rewind2 (x: xs) l = rewind2 xs (x: l)
```

Comment:
As you correctly noticed, that implemetation with `++ [x]` will be slow, and the manual implementation is much better :top: 

It is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

The helper function in such cases is usually called `go`. You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

Reply:
Thank you for teaching me HASKELL MINI-PATTERNS HANDBOOK. It looks very useful. I bookmarked it and will read later.